### DISCUSSION TITLE
Callsite Revalidation Opt-out

### DISCUSSION THREAD

**Discussion by ryanflorence** [remix-run/react-router#10006] [Proposals]
## tl;dr

```tsx
<Form shouldRevalidate={(revalidateFuncArg) => false} />
fetcher.submit(data, { shouldRevalidate: () => true });
```

## Background

After actions are called, for any reason, a revalidation is automatically initiated. You can opt-out of that revalidation with the route `shouldRevalidate` property. This works great for a lot of cases and is familiar to geriatric React developers like me who remember `shouldComponentUpdate`.

One way to think about our current “revalidation opt-out” is that it is part of a routes “definition”, or a “definition site opt-out”. It doesn’t matter what triggered the revalidation, the route’s definition gets to decide if it wants to participate in revalidation. This is good because the idea is that developers can look at what the `loader` uses in order to define the route’s `shouldRevalidate`. If it uses a certain search param, or url params, or loads data related to some special formData going to the action, all you have to know is the route’s own loader and not worry about the rest of the app.

However, sometimes these `shouldRevalidate` functions can be tricky write with that model, and developers (including myself) often over-optimize (causing stale data bugs) or under-optimize (making it pointless).

I’ve found in my own code that it’s often easier to answer the question “should this route revalidate?” at the “call site”: the place where a `<Form>` is rendered or a `fetcher.submit()` is called.

## Use Case:

A very easy case to think about is a newsletter sign up form at the bottom of multiple pages. If three routes are matching on a page then three revalidation requests will go out when the user submits the form. But, being subscribed to the newsletter is irrelevant to every loader in the app. 

To get Remix to avoid making all three revalidation fetches for the newsletter form, each route must define a `shouldRevalidate`. Typically it’s far easier to check if the `formData` is for the newsletter signup instead of what the loader needs. This is not the “definition site” model (what does this loader care about?) but is instead whack-a-mole for every route for every action that is irrelevant to loaders.

## Open questions:

- GET revalidation (URLSearchParam changes), `<Link> <Form> fetcher.load`

**Comment by AndreSilva1993** [comment]
Are there any news on this one? I'm currently handling a sort of similar scenario where I'm using a `fetcher.Form` to dismiss notifications and would like to avoid having every loader revalidated - since the notifications are being fetched via a resource route periodically, there's no need at all to update any information present in the page.

Currently I'm thinking of just opting out entirely off the `useFetcher` API and just using a plain fetch plus `useState` to have optimistic UI - whereas I would much rather prefer the `useFetcher / fetcher.formData` combo since it works flawlessly, apart from the automatic revalidations.

**Reply by codeAdityagithub** [reply]
Is there any update? I also stumbled upon this when i was building a comment section. The replies for a comment are fetched when the user clicks to see them, and the replies are added to that component only. But when I use fetcher.submit to submit the reply, remix revalidates all the above comments also that user has fetched replies for. I am rather new to remix and i am not able to find a way to handle this.

**Comment by willhoney7** [comment]
This is our biggest pain point for Remix. In addition to callsite revalidation opt-out, we would greatly benefit from callsite revalidation opt-_in_. `useRevalidator` is great, but can I please just revalidate a single route? I think I'm going to end up creating a "revalidate data X" action (which will be a noop on the backend), and then create a global `shouldRevalidate` for every single other route in our app that knows to ignore that one. Woof!

**Comment by rishitells** [comment]
> This is our biggest pain point for Remix. In addition to callsite revalidation opt-out, we would greatly benefit from callsite revalidation opt-in. useRevalidator is great, but can I please just revalidate a single route? I think I'm going to end up creating a "revalidate data X" action (which will be a noop on the backend), and then create a global shouldRevalidate for every single other route in our app that knows to ignore that one. Woof!

+1.

We really need callsite revalidation opt-out for some of our use cases. 
@ryanflorence @brophdawg11 Are there any plans on this soon(ish)? 😅

**Comment by okalil** [comment]
Wouldn't be simpler a boolean prop, maybe `preventRevalidation` (similar to `preventScrollReset`)? What would be the use case of shouldRevalidate as a function?

**Comment by timfrans** [comment]
We're having a similar use case for this as @AndreSilva1993.

We have a resource route defined and want the loader of this route only on specific occasions.

**Comment by dadamssg** [comment]
adding a +1.

I have a "tags" form component where this is problem. Think github labels where they can be created on the fly. These tags can be applied to all sort of things throughout the app. Having to add, and remember to add, `shouldRevalidate` functions to every route/subroute where this component can be used is incredibly tedious and prone to error.

**Comment by nokwandapatiencechiya-png** [comment]
What You get  if you realize code

**Comment by dadamssg** [comment]
I took a crack at this: https://github.com/remix-run/react-router/pull/14542

**Comment by brophdawg11** [comment]
A good point was raised in discord that this should maybe apply to GET navigations as well - primarily so you can set search params triggering loaders, but also since the default in single fetch is to revalidate - that would allow you to perform normal navigations and avoid revalidating layout routes (at the expense of more cache key enumerations due to `.data` query params).

**Reply by brophdawg11** [reply]
Just noticed this is already captured above in "Open questions" 🙃

**Comment by brophdawg11** [comment]
We talked about this in the steering committee meeting today and going to move forward with it with some minor alterations from the original proposal.  We agree it makes sense to include this on navigations as well to allow for opting out of revalidations on setting certain search params etc.

```tsx
fetcher.submit(data, { defaultShouldRevalidate: false }); ✅
navigate("/", { defaultShouldRevalidate: false }); ✅
submit(data, { method: 'post', action: '/', defaultShouldRevalidate: false }); ✅
```

In order to allow routes to be the final decision maker on revalidation, if a route exports a `shouldRevalidate`, it will be called with the provided value for `defaultShouldRevalidate` to determine behavior.  

If no active routes have a `shouldRevalidate`, then the provided value will be used directly.

We **don't plan** to add the function format - the value you provide should be computed off local values, n ot us proxying through the same params for route-level `shouldRevalidate`.

```tsx
<Form shouldRevalidate={(revalidateFuncArg) => false} /> ❌
```

**Reply by brophdawg11** [reply]
@dadamssg Do you want to take a stab at updating your PR and seeing if you can flow that through to the `defaultShouldRevalidate` section via `getMatchesToLoad`?  Feel free to take a stab at the navigation variants too, but if that's more than you were looking to take on I'm happy to handle that side of it.  Let me know if you have any questions!

**Reply by dadamssg** [reply]
@brophdawg11 yes, i'll try to rework the PR when i get a chance. glad to see movement on this!

**Comment by rossipedia** [comment]
IIUC, this proposal should address #9851 as well, right? Assuming `defaultShouldRevalidate` is added as a prop to `<Link>` and an option to `setSearchParams` as well?

**Reply by brophdawg11** [reply]
That issue has become like 3 separate conversations (skip re-rendering, skip revalidation, skip scroll reset), but I don't think the original issue (prevent re-renders) is covered by this.  The `location` is managed by a top-level context so when you navigate that context (specifically `router.state.location`) will update and cause re-renders.

We talked about that a while back and we;'re not interested in trying to avoid re-renders on context changes - that's basically the ethos of React.  We're not trying to achieve fine-grained reactivity within the router.

**Reply by rossipedia** [reply]
Ahh ok. Yeah I was thinking it was a bit of an XY problem, because re-renders are almost never the real pain point, it's always that a re-render causes something like a network request or some expensive IO or something like that.

**Comment by brophdawg11** [comment]
This is available in `0.0.0-experimental-2d8f5c807` if anyone wants to do some alpha testing!  Please let us know if you run into any issues.

Thanks to @dadamssg for the [PR](https://github.com/remix-run/react-router/pull/14542)!

**Reply by brophdawg11** [reply]
Updated experimental release: `0.0.0-experimental-112589a91`

**Comment by brophdawg11** [comment]
stabilized in 7.15.0
